certora/specs/CER-2-Operator/CER-68-Operator-authorization.spec

```diff
@@ -0,0 +1,147 @@
+    // Assume: In the inductive step, for the precondition,
+    // we are assuming that the low-level `CALL`s that are
+    // reachable in batch/call will not affect the state of ownerLookup
+    ownerLookupGhost[prefix] == 0 || getAddressPrefix(ownerLookupGhost[prefix]) == prefix
+    filtered { 
+        f -> !isMustRevertFunction(f)
+    }
+
+/**
+  * Checks a liveness property that the owner of an account
+  * can succesfully set an operator (under a few assumptions
+  * that are spelled out with the "require" statements).
+ */
+rule theOwnerCanCallSetOperator() {
+    env e;
+
+    bytes19 addressPrefix;
+    address operator;
+    uint256 operatorBitField;
+
+    address owner = getOwnerOf(addressPrefix);
+    requireInvariant OwnerIsFromPrefix(addressPrefix);
+
+    // the actual caller (either msg.sender or the onBehalfOfAccount)
+    address caller = actualCaller(e);
+
+    // This is a permit self-call:
+    if (e.msg.sender == currentContract) {
+        // we are within permit() and should use getOnBehalfOfAccount() instead
+        caller = getExecutionContextOnBehalfOfAccount(e);
+        // the owner is already set, not zero and not EVC
+        require(owner == caller && owner != 0 && owner != currentContract);
+        // the owner has the proper prefix
+        require(getAddressPrefix(owner) == addressPrefix);
+    // This is the normal case where the caller is msg.sender:
+    } else {
+        // just a regular call from msg.sender
+        // msg.sender is from the proper prefix
+        require(getAddressPrefix(e.msg.sender) == addressPrefix);
+        // the owner is not set yet (zero) or identical to msg.sender
+        require(owner == 0 || owner == e.msg.sender);
+    }
+
+    // The function will revert if any of these assumptions do not hold:
+    require(e.msg.value < nativeBalances[e.msg.sender]);
+    require operator != 0;
+    require !(operator == currentContract);
+    require !(haveCommonOwner(caller, operator));
+
+    // the operator can not be from the prefix either
+    require(getAddressPrefix(operator) != getAddressPrefix(caller));
+
+    // the current bitfield must be different from what we try to set
+    require(getOperator(addressPrefix, operator) != operatorBitField);
+
+    // call the setOperator() method.
+    setOperator@withrevert(e, addressPrefix, operator, operatorBitField);
+    // check that it does not revert under these assumptions
+    assert(!lastReverted);
+}
+
+// Only the owner of an account can set the operator of that account
+// if it is called with authenticated=true
+rule onlyOwnerOrOperatorCanCallSetAccountOperator() {
+    env e;
+    address account;
+    address operator;
+
+    address caller = actualCaller(e);
+
+    // call the setAccountOperator method.
+    setAccountOperator(e, account, operator, true);
+
+    address owner = haveCommonOwner(account, caller) ? caller : getAccountOwner(e, account);
+
+    // Since setAccountOperator did not revert, the actualCaller
```

Comment:
typo: `Since setAccountOperator did not revert, the actualCaller must be the owner`

Reply:
fixed